Di-container translated into Russian (#204)
yiisoft/docs: guide/ru/concept/di-container.md

# Внедрение зависимостей и контейнер внедрения зависимостей

## Внедрение зависимостей <span id="dependency-injection"></span>

> В ООП существует два способа повторного использования кода: наследование и  композиция.

Reviewer: ```suggestion
В ООП существует два способа повторного использования кода: наследование и композиция.
```

Author: Fixed

Наследование - это просто:

```php
class Cache
{
    public function getCachedValue($key)
    {
        // ..
    }
}

class CachedWidget extends Cache
{
    public function render(): string
    {
        $output = $this->getCachedValue('cachedWidget');
        if ($output !== null) {
            return $output;
        }
        // ...        
    }
}
```

> Проблема здесь в том, что эти два класса становятся излишне сопряженными или взаимозависимыми, что делает их более хрупкими.

Reviewer: ```suggestion
Проблема здесь в том, что эти два класса становятся излишне связанными или взаимозависимыми, что делает их более хрупкими.
```

Author: Я ориентировался на перевод книги "Совершенный код" Стива Макконнела: coupling - сопряжение, cohesion - связность.

> Понятие связности (cohesion) возникло в области структурного проектирования
> и обычно обсуждается в том же контексте, что и сопряжение (coupling).

Reviewer: Хорошо, пусть сопряжение остаётся. Не критично. С переводом cohesion/coupling всегда всё неоднозначно.

Author: Понял :)

Есть способ справиться с этой проблемой — композиция:

```php
interface CacheInterface
{
    public function getCachedValue($key);
}

final class Cache implements CacheInterface
{
    public function getCachedValue($key)
    {
        // ..
    }
}

final class CachedWidget
{
    private CacheInterface $cache;

    public function __construct(CacheInterface $cache)
    {
        $this->cache = $cache;
    }
    
    public function render(): string
    {
        $output = $this->cache->getCachedValue('cachedWidget');
        if ($output !== null) {
            return $output;
        }
        // ...        
    }
}
```

> Мы избежали ненужного наследования и использовали интерфейс, чтобы уменьшить сопряженность.

Reviewer: ```suggestion
Мы избежали ненужного наследования и использовали интерфейс, чтобы уменьшить связанность.
```

```php
use \Yiisoft\Cache\CacheInterface;

class MyController
{
    private CacheInterface $cache;

    public function __construct(CacheInterface $cache) {
        $this->cache = $cache;    
    }

    public function actionDashboard(RevenueReport $report)
    {
        $reportData = $this->cache->getOrSet('revenue_report', function() use ($report) {
            return $report->getData();               
        });

        return $this->render('dashboard', [
           'reportData' => $reportData,
        ]);
    }
}
```

Поскольку именно [yiisoft/injector](https://github.com/yiisoft/injector) создает экземпляр и вызывает обработчик действий - он проверяет типы аргументов конструктора и метода, получает зависимости этих типов из контейнера и передает их как аргументы.
Обычно это называется автоматическим разрешением зависимостей.
Это происходит и с дополнительными зависимостями - если вы явно не указываете зависимость, контейнер сначала проверит, есть ли у него такая зависимость.
Достаточно объявить нужную вам зависимость, и она будет получена из контейнера автоматически.

## Полезные ссылки <span id="references"></span>

> - [Inversion of Control Containers and the Dependency Injection pattern Мартин Фаулер](https://martinfowler.com/articles/injection.html)

Reviewer: ```suggestion
- [Inversion of Control Containers and the Dependency Injection pattern Мартина Фаулера](https://martinfowler.com/articles/injection.html)
```

Author: Fixed
